dll/win32/shell32/shlexec.cpp

```diff
@@ -458,6 +458,132 @@ static HRESULT SHELL_GetPathFromIDListForExecuteW(LPCITEMIDLIST pidl, LPWSTR psz
+                    wsprintfW(fmtbuf, buf, L"(%s)"); // Change "Blah blah %s" to "Blah blah (%s)"
+                    wsprintfW(buf, fmtbuf, pData->NameBuffer);
+                    SendMessage(control, WM_SETTEXT, 0, (LPARAM)buf);
+                    SendDlgItemMessage(hwnd, IDC_RUNAS_NAME, CB_ADDSTRING, 0, (LPARAM)pData->NameBuffer);
+                }
+                SendDlgItemMessage(hwnd, IDC_RUNAS_NAME, CB_SETCURSEL, 0, 0);
+                return TRUE;
+            }
+            break;
+        case WM_COMMAND:
+            switch(LOWORD(wParam))
+            {
+                case IDCANCEL:
+                    EndDialog(hwnd, 1);
+                    break;
+                case IDOK:
+                    {
+                        pData->Safer = SendDlgItemMessage(hwnd, IDC_RUNAS_SAFER,
+                                                          BM_GETCHECK, 0, 0);
+                        SendDlgItemMessage(hwnd, IDC_RUNAS_NAME, WM_GETTEXT,
+                                           _countof(RUNASDLGDATA::NameBuffer), (LPARAM)pData->NameBuffer);
+                        SendDlgItemMessage(hwnd, IDC_RUNAS_PASS, WM_GETTEXT,
+                                           _countof(RUNASDLGDATA::Pass), (LPARAM)pData->Pass);
+                        UINT other = SendDlgItemMessage(hwnd, IDC_RUNAS_OTHER,
+                                                        BM_GETCHECK, 0, 0) != 0;
+                        pData->Name = pData->NameBuffer;
+                        pData->Domain = strstrW(pData->Name, L"\\");
+                        if (pData->Domain)
+                        {
+                            LPWSTR tmp = pData->Domain + 1;
+                            pData->Domain[0] = UNICODE_NULL;
+                            pData->Domain = pData->Name;
+                            pData->Name = tmp;
+                        }
+                        if (other)
+                        {
+                            HANDLE token;
+                            if (!LogonUser(pData->Name, pData->Domain, pData->Pass,
+                                           LOGON32_LOGON_INTERACTIVE, LOGON32_PROVIDER_DEFAULT,
+                                           &token))
+                                return SHELL_ErrorBox(hwnd);
+                            CloseHandle(token);
+                            other |= (GetKeyState(VK_SHIFT) & 0x8000);
+                        }
+                        EndDialog(hwnd, 2 + other);
+                    }
+                    break;
+                case IDC_RUNAS_BROWSE:
+                    return SHELL_ErrorBox(hwnd, ERROR_CALL_NOT_IMPLEMENTED); // TODO
+            }
+            break;
+    }
+    return FALSE;
+}
+
+static HRESULT SHELL_PromptAndRunProcessAs(HWND hwnd, LPWSTR Cmd, DWORD CreationFlags,
+                                           LPWSTR Env, LPCWSTR Dir,
+                                           STARTUPINFOW *pSI, PROCESS_INFORMATION *pPI)
```

Comment:
New function (and won't be synced to wine), so you can use more proper formatting:
```suggestion
static
HRESULT
SHELL_PromptAndRunProcessAs(
    _In_ HWND hwnd,
    _In_ LPWSTR Cmd,
    _In_ DWORD CreationFlags,
    _In_opt_ LPWSTR Env,
    _In_opt_ LPCWSTR Dir,
    _In_ LPSTARTUPINFOW pSI,
    _Out_ LPPROCESS_INFORMATION pPI)
```